### DISCUSSION TITLE
Postprocessor for arbitrary linear equation

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#25616] [Q&A General]
I'd like to do something like a NodalSum Postprocessor, except I'd like the coefficients to be artibrary, ie. Ax + By + Cz +.... It would be boundary-restricted, so in theory the coefficients would correspond to the order of the nodes in my nodeset. I'm thinking of inhereting from the NodalSum PPS and overriding the `execute` method like so:

```
void
NodalSum::execute()
{
  _sum += _coef * _u[_qp];
}
```

I'm just not sure how to make sure I'm getting the right coeff for the particular node. Any ideas?

**Comment by GiudGiud** [comment]
If it s as simple as Ax + By + Cz you can use a function for the coefficient, and do this:

```
_sum += _coef.value(_t, *_current_node);
```

**Reply by jessecarterMOOSE** [reply]
Sorry, maybe I should be more specific. The x, y, and z are _u at different nodes. The equation reads more like:

$A_0 u_0 + A_1 u_1 + A_2 u_2 + ...$

Where $A_i$'s are coefficients and $u_i$'s are my variables at nodes in my nodeset

**Reply by GiudGiud** [reply]
ah ok then what does A depend on?

**Reply by jessecarterMOOSE** [reply]
It's a constant coefficient particular to the node.

**Reply by jessecarterMOOSE** [reply]
So each node in the node set gets a different coefficient.

**Reply by GiudGiud** [reply]
where do you know them?
are they in a file?

**Reply by jessecarterMOOSE** [reply]
For my case there will only be just a few, so I'm thinking about inputting them as parameters in the input file

**Reply by GiudGiud** [reply]
I think a function makes sense here. And it has plenty of flexibility for the input.

you can bake in a space dependence fairly easily.

what does that variation depend on? Are they different coefficients per subdomain?

**Reply by jessecarterMOOSE** [reply]
I'm thinking if, since the pps will be boundary-restricted, getting the current node id (`_current_node->id()`), find its position in the nodeset, then apply the coefficient at the same position in the input vector of values. Does that make sense?

**Reply by jessecarterMOOSE** [reply]
No variation in the coefficients, they would be constant in space and time.

**Reply by GiudGiud** [reply]
I don’t think nodes are ordered within a node set. How are you ordering them?

If the coefficient depends on the node, it depends on space right?

**Reply by jessecarterMOOSE** [reply]
> I don’t think nodes are ordered within a node set. How are you ordering them?

Ah I had assumed they were

> If the coefficient depends on the node, it depends on space right?

True, though I'm given the node numbers and coefficients in a separate file, so I would need to do some kind of lookup on node coordinates as intermediate step

**Reply by GiudGiud** [reply]
so the PropertyReadFile will read data by node id.

You will need to "fill in" with 0s to make sure the range of node id is completely covered.
But then a PiecewiseConstrantFromCSV with the node read_type option will match the node id to a CSV data

**Reply by jessecarterMOOSE** [reply]
Could there be an easier way? Perhaps a PPS is not what I need?

**Reply by GiudGiud** [reply]
If you want a single scalar value, a PP is a good idea.
A NodalPP will have the boundary-restricted nodal loop you are looking for too. It s not very hard to build from scratch in a GeneralUserObject if you prefer hacking

**Reply by jessecarterMOOSE** [reply]
I do like the function idea, is there anything like a Dirac delta function in Moose?

**Reply by GiudGiud** [reply]
Dont think so.
You d still be reading a file for the coordinates or for the node ids to know where to put it non-zero values?

**Reply by jessecarterMOOSE** [reply]
Yeah, something like that

**Reply by GiudGiud** [reply]
You could just modify the PropertyFileReader to handle node ids in the way you like

**Reply by jessecarterMOOSE** [reply]
What about a PPS that evaluates a parsed expression at a node? So I would have these parameters:
```
node = 123
expression = 'a*u + b*v'
variables = 'u, v'
constants = '1, 2'
```
Is there a postprocessor that can do something like that?

**Reply by GiudGiud** [reply]
If you add a nodeset that only includes node 123 you can use all the SidePP-derived postprocessors only on one node. 

How many nodes do you have that you want to look at?

**Reply by jessecarterMOOSE** [reply]
> How many nodes do you have that you want to look at?

Just a few, number would be in the single digits

**Reply by GiudGiud** [reply]
Sideset for each is fine then?

**Reply by jessecarterMOOSE** [reply]
Yes that would be reasonable. Is there are particular PP you have in mind?

**Reply by GiudGiud** [reply]
btw there's also the PointValue type postprocessors you could adapt if you have that few. I ve seen that used quite a bit

**Reply by GiudGiud** [reply]
EDIT not SidePP, NodalPP!

NodalMaxValue on a single node nodeset gives you u for example

**Reply by jessecarterMOOSE** [reply]
OK so I could do that for each term and then combine with ParsedPostprocessor

**Reply by GiudGiud** [reply]
yeah for example.
Or write you own NodalPP

A NodalFunctorAveragePP for example would be welcome
